### PR TITLE
split fondant component into load and transform

### PR DESCRIPTION
**Pull request by PhilippeMoussalli**
PR that splits the Fondant Component into load and transform component. The user will now call `FondantLoaderComponent` or `FondantTransoromComponent` depending on the type of job that is run. 

A few notes: 

- Both components rely on the `FondantComponentSpec` which requires both `input_subsets` and defines an `input_manifest_path` as required argument (based on `KubeflowComponentSpec`). In practice, the loader component should not require both inputs. Wondering if it's worth splitting the Spec to loader and transform as well and initialize both `FondantLoaderComponent` and  `FondantTransformComponent` with those. 

- Currently only tested for the `FondantLoaderComponent`. I think the `FondantTransformComponent` still needs to be implemented but part of another PR (todos are still present in the code)

### CODE REVIEW

File: examples/pipelines/simple_pipeline/components/load_from_hub/src/main.py [hunk 1]
```diff
@@ -11,7 +11,7 @@
 import numpy as np
 from PIL import Image
 
-from fondant.dataset import FondantComponent
+from fondant.dataset import FondantLoaderComponent
```
Comment (NielsRogge) [inline examples/pipelines/simple_pipeline/components/load_from_hub/src/main.py:14]:
```suggestion
from fondant.dataset import FondantLoadComponent
```
Let's call them "load" and "transform"

File: fondant/dataset.py [hunk 2]
```diff
@@ -166,82 +167,123 @@ def upload(self, save_path):
...
-            df = dataset.load_data(self.spec)
-            # provide this dataframe to the user
-            df = self.transform(
-                df=df,
-                args=args,
-            )
-            # TODO update index, potentially add new subsets
+        component_arguments = {}
+        kubeflow_component_spec = self.spec.kubeflow_specification
+        component_arguments.update(kubeflow_component_spec.input_arguments)
```
Comment (NielsRogge) [inline fondant/dataset.py:192]: just to confirm, the custom `args` of a component are part of the `input_arguments`?

Comment (PhilippeMoussalli) [reply fondant/dataset.py:192]: yes indeed the kubeflow input arguments are derived from the ComponentSpec of the component which contains the arguments

File: examples/pipelines/simple_pipeline/components/load_from_hub/src/main.py (hunk 1 above)
Comment (PhilippeMoussalli) [reply examples/pipelines/simple_pipeline/components/load_from_hub/src/main.py]: done :)

File: fondant/dataset.py [hunk 3]
```diff
@@ -166,82 +167,123 @@ def upload(self, save_path):
...
+    @abstractmethod
+    def load(self, args: argparse.Namespace) -> dd.DataFrame:
+        """Abstract method that loads the initial dataframe"""
+
+    def run(self):
+        """
+        Runs the loading component.
+        """
+        # step 1: Add and parse arguments
+        args = self._add_and_parse_args()
```
Comment (RobbeSneyders) [inline fondant/dataset.py:253]: Can we lift these common steps to the base class?

File: fondant/dataset.py [hunk 4]
```diff
@@ -166,82 +167,123 @@ def upload(self, save_path):
...
+
+    def __init__(self):
         self.spec = FondantComponentSpec.from_file("fondant_component.yaml")
-        self.type = type
 
-    def run(self) -> dd.DataFrame:
-        """
-        Parses input data, executes the transform, and creates the output manifest.
+    @abstractmethod
+    def _get_manifest(self, args: argparse.Namespace) -> Manifest:
```
Comment (RobbeSneyders) [inline fondant/dataset.py:177]: Not a fan of defining abstract methods which are not called by the base class. It imposes a structure on the subclasses without any advantage to them.

File: fondant/dataset.py [hunk 5]
```diff
@@ -166,82 +167,123 @@ def upload(self, save_path):
...
+
+    @abstractmethod
+    def load(self, args: argparse.Namespace) -> dd.DataFrame:
+        """Abstract method that loads the initial dataframe"""
+
+    def run(self):
+        """
+        Runs the loading component.
+        """
+        # step 1: Add and parse arguments
```
Comment (RobbeSneyders) [inline fondant/dataset.py:252]: These comments are a bit redundant here. The code is quite self explanatory.

File: fondant/dataset.py [hunk 6]
```diff
@@ -166,82 +167,123 @@ def upload(self, save_path):
...
+    """Abstract base class for a Fondant loader component"""
+
+    def _add_and_parse_args(self) -> argparse.Namespace:
         parser = argparse.ArgumentParser()
+        component_arguments = self.get_component_arguments()
 
-        kubeflow_component = self.spec.kubeflow_specification
+        for arg in component_arguments.values():
+            # Input manifest is not required for loading component
+            if arg.name == "input_manifest_path":
```
Comment (RobbeSneyders) [inline fondant/dataset.py:210]: I would expect this arg to not be defined for a Load component instead of having to handle it specifically here. Without this, we can lift this method completely to the base class.

**Review by RobbeSneyders — COMMENTED** [review]
Thanks @PhilippeMoussalli!

File: fondant/dataset.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (RobbeSneyders) [reply fondant/dataset.py:177]: As mentioned below, I think lifting the code that calls this method to the base class is the best solution.

File: fondant/dataset.py [hunk 7]
```diff
@@ -166,82 +167,123 @@ def upload(self, save_path):
...
+        parser.add_argument(
+            "--metadata",
+            type=str,
+            required=True,
+            help="The metadata associated with the pipeline run",
+        )
+
+        return parser.parse_args()
+
+    def _get_manifest(self, args: argparse.Namespace) -> Manifest:
```
Comment (NielsRogge) [inline fondant/dataset.py:232]:
```suggestion
    def _create_manifest(self, args: argparse.Namespace) -> Manifest:
```

File: fondant/dataset.py (hunk 6 above)
Comment (PhilippeMoussalli) [reply fondant/dataset.py:210]: Fair point, can we just then keep it as Optional by default for both components. Otherwise we might need to rework the `KubeflowComponentSpec` to have different specification for both loader and transform. https://github.com/ml6team/fondant/blob/main/fondant/component_spec.py#:~:text=class-,KubeflowComponentSpec,-%3A
@NielsRogge wondering what's your take on this?

Comment (NielsRogge) [reply fondant/dataset.py:210]: I feel like the user would only need to be exposed to a single `FondantComponent` class actually. We could figure out that the first component in the pipeline is the loading one and that all others are transform components. We can internally split up the kubeflow spec for loading vs transformation, indeed.

Comment (PhilippeMoussalli) [reply fondant/dataset.py:210]: Good point. I think we can eventually tackle but maybe in a separate PR.

File: fondant/dataset.py (hunk 7 above)
Comment (PhilippeMoussalli) [reply fondant/dataset.py:232]: This is an abstract function that might either get the manifest if it exists or create a new one 
perhaps `load_or_create_manifest` might be a better name

File: fondant/dataset.py (hunk 6 above)
Comment (RobbeSneyders) [reply fondant/dataset.py:210]: The interface for both components is different, since we need to provide an input dataframe for the transform components and not for the load components. I don't think it makes sense to try to align these completely.

File: fondant/dataset.py (hunk 5 above)
Comment (NielsRogge) [reply fondant/dataset.py:252]: The comments below are helpful to me

File: fondant/dataset.py (hunk 6 above)
Comment (NielsRogge) [reply fondant/dataset.py:210]: Ok, we can keep the separation also for the user in that case.

File: fondant/dataset.py (hunk 7 above)
Comment (NielsRogge) [reply fondant/dataset.py:232]: Ok, feel free to rename

File: fondant/dataset.py (hunk 5 above)
Comment (RobbeSneyders) [reply fondant/dataset.py:252]: I really don't see the value of eg. these comments as the same information is clear from the lines of code. Let's only add inline comments when they provide additional information on the reasoning behind a piece of code or when the logic is too hard to follow in code.

```
# step 2: Get manifest
manifest = self._get_manifest(args)

# step 3: Create dataset
dataset = FondantDataset(manifest)
```

Comment (NielsRogge) [reply fondant/dataset.py:252]: Yes I meant the comments of the steps below step 3

File: fondant/dataset.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (PhilippeMoussalli) [reply fondant/dataset.py]: Thanks for the suggestion. Indeed it's much nicer to enforce and define where those abstract methods should be used in the base class. I updated the PR accordingly

File: fondant/dataset.py (hunk 3 above)
Comment (PhilippeMoussalli) [reply fondant/dataset.py]: Done :)

File: fondant/dataset.py (hunk 5 above)
Comment (PhilippeMoussalli) [reply fondant/dataset.py]: I removed some of the redundant comments and left the ones that could be useful in providing more clarity

**Review by RobbeSneyders — APPROVED** [review]
Thanks @PhilippeMoussalli!
